examples/go/fakemodem/client.go

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Infostellar, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+
+	api "github.com/infostellarinc/go-stellarstation/api/v1/groundstation"
+)
+
+/**************
+ * API Client *
+ **************/
+
+// Client is a StellarStation Ground Station API client.
+type Client struct {
+	groundstation GroundStationConfig
+	client        api.GroundStationServiceClient
+	runner        *Runner
+}
+
+// NewClient creates a new Client instance
+func NewClient() *Client {
+	client := &Client{
+		runner: NewRunner(),
+	}
+	return client
+}
+
+// Connect connects to the ground station.
+func (c *Client) Connect(groundstation GroundStationConfig) {
+	startFunction := func(ctx context.Context) {
+		log.Printf("Connecting to ground station %v (%v).\n", groundstation.Name, groundstation.ID)
+		log.Printf("Ground station configuration: %+v\n", groundstation)
+
+		c.groundstation = groundstation
+
+		c.connect()
+	}
+
+	stopFunction := func() {
+		log.Printf("Disconnecting from ground station %v (%v).\n", groundstation.Name, groundstation.ID)
+	}
+
+	c.runner.Start(startFunction, stopFunction)
```

Comment:
I may miss something, but I think Client.Connect doesn't have to be run with Runner. Just simply writing connection logic in the startFunction and post logging in the stopFunction makes the logic simpler and testable.

Reply:
The runner is used to close out the grpc connection when either the program exits or a second call to `Connect()` is made.  If we change the code so that the client can not be re-connected to a second star pass (which seems logical) then the code can be made simpler by maintaining a reference to the connection and using a mutex to set/retrieve it.